File: scripts/lib/prompts.py
"""Prompt template loading and substitution.

Templates use string.Template ($var) rather than str.format — the templates
talk about GIFT and JSON, so literal braces are frequent and must survive.
Validation happens on the TEMPLATE, not on the rendered text: substituted
content (a transcript, a session prompt) may legitimately contain $-words,
so scanning the output would false-positive. Instead the template's
identifiers must exactly be covered by the provided variables (missing one
raises; an unused variable only warns — it may be intentional per-template).
"""

from __future__ import annotations

import sys
from pathlib import Path
from string import Template


class PromptError(Exception):
    pass


def load_template(path: Path) -> str:
    if not path.exists():
        raise PromptError(f"prompt template not found: {path}")
    return path.read_text(encoding="utf-8")


def _identifiers(tmpl: Template, text: str) -> set[str]:
    ids: set[str] = set()
    for match in tmpl.pattern.finditer(text):
        if match.group("invalid") is not None:
            raise PromptError(f"malformed '$' placeholder in template near: {text[match.start():match.start() + 30]!r}")
        name = match.group("named") or match.group("braced")
        if name:
            ids.add(name)
    return ids


def render(template_text: str, variables: dict[str, object]) -> str:
    tmpl = Template(template_text)
    ids = _identifiers(tmpl, template_text)
    missing = ids - variables.keys()
    if missing:
        raise PromptError(f"prompt template references {sorted('$' + m for m in missing)} but no value was provided")
    unused = variables.keys() - ids
    if unused:
        print(f"⚠️  prompt variables provided but not used by template: {sorted(unused)}", file=sys.stderr)
    return tmpl.substitute({k: str(v) for k, v in variables.items()})

File: app/core/prompts.js
// Prompt template loading and substitution — JS port of scripts/lib/prompts.py.
// Same contract as Python string.Template: $var / ${var}, $$ escapes a dollar,
// any other '$' is malformed. Validation happens on the TEMPLATE, not the
// rendered text (substituted content may legitimately contain $-words).

export class PromptError extends Error {}

const PATTERN = /\$(?:(\$)|([_a-zA-Z][_a-zA-Z0-9]*)|\{([_a-zA-Z][_a-zA-Z0-9]*)\})/g;

function identifiers(text) {
  const ids = new Set();
  let last = 0;
  for (const m of text.matchAll(PATTERN)) {
    checkInvalid(text, last, m.index);
    last = m.index + m[0].length;
    const name = m[2] || m[3];
    if (name) ids.add(name);
  }
  checkInvalid(text, last, text.length);
  return ids;
}

// A '$' not consumed by the pattern is Python's "invalid" group.
function checkInvalid(text, from, to) {
  const i = text.slice(from, to).indexOf("$");
  if (i !== -1) {
    const at = from + i;
    throw new PromptError(`malformed '$' placeholder in template near: ${JSON.stringify(text.slice(at, at + 30))}`);
  }
}

export function render(templateText, variables) {
  const ids = identifiers(templateText);
  const provided = new Set(Object.keys(variables));
  const missing = [...ids].filter((k) => !provided.has(k)).sort();
  if (missing.length) throw new PromptError(`prompt template references ${JSON.stringify(missing.map((m) => "$" + m))} but no value was provided`);
  const unused = [...provided].filter((k) => !ids.has(k)).sort();
  const warnings = unused.length ? [`⚠️ prompt variables provided but not used by template: ${JSON.stringify(unused)}`] : [];
  const out = templateText.replace(PATTERN, (whole, dollar, named, braced) => (dollar ? "$" : String(variables[named || braced])));
  return { text: out, warnings };
}
